README.md
# BalenaFin Co-Processor Firmata

This is an implementation of the [Firmata](https://github.com/firmata/protocol) protocol for Silicon Labs BGM111. It is compatible with standard Firmata 2.5.8. 

## Installation

### Dockerfile

The easiest way to install the Firmata application onto your board is to run the [balena application](https://github.com/balena-io/balena-fin-firmata-flash) provided. This targets the latest verision of the Balena Firmata application. 
This balena application will run and install [OpenOCD](http://openocd.org/) on your Fin in order to provision the Coprocessor with both a bootloader and the Firmata application.

We also provide a [Dockerfile](Dockerfile) for you to easily generate an environment for compiling this firmware yourself.

#### Build

To build the docker image, run:

```bash
docker build . --tag firmata 
```

A container can then be run with:

```bash
docker run -it -v ${YOUR-OUTPUT-DIRECTORY}:/out --name firmata-build firmata "make" "all"
```

> :wrench: * `all` can be substituted for `balena` if you only wish to target to balenaFin or `devkit` for the Silabs BRD4001A

This will output the build files to a directory specified by `YOUR-OUTPUT-DIRECTORY`, e.g. `~/Downloads/balena`.
The output binaries for flashing the coprocessor can be found under `${YOUR-OUTPUT-DIRECTORY}/builds/balena`.

If you wish to stop this container:

`docker stop firmata-build`

And to remove the container:

`docker rm firmata-build`

### Manually Build

It is also possible to build the source and manually flash the Coprocessor however, in order to flash the Coprocessor you will need to either load the compiled firmware onto the Compute Module and flash it using OpenOCD or program the Coprocessor using an external programmer such as a [Segger JLink ](https://www.segger.com/products/debug-probes/j-link/).

#### Dependencies

Before attempting to build, ensure the following dependencies are installed. 
We recommend ARM GCC version `9-2019-q4-major`, as used in our docker image.

 - cmake
 - make
 - arm-none-eabi-gcc

> :wrench: Make sure these are all located in your `$PATH`

#### Build

With the dependencies installed, build with the following commands:

1. `cd src` to change directory to the source files
2. `make setup` to generate the builds directory
3. `make balena` to execute the build*

> :wrench: * `make devkit` can be used to build for the Silabs BRD4001A

### Flashing

As the coprocessor is coupled to the compute module, we recommend that you use our flashing [example](https://github.com/balena-io/balena-fin-firmata-flash) to automate the flashing process.

#### Flashing via balenaFin Compute Module

If you wish to deploy this coprocessor firmware to a balenaFin running either [balenaOS](https://dashboard.balena-cloud.com) or [Raspbian](https://github.com/balena-os/pi-gen/releases), checkout this [example](https://github.com/balena-io/balena-fin-firmata-flash).

#### JLink Programmer

An external JLink programmer may be used to flash the coprocessor (for example, if using a Silabs Development Kit)

 - [JLink tools](https://www.segger.com/jlink-software.html)

To flash, using an external JLink programmer:

1. `cd builds/devkit && make flash` to flash to a device

## Firmata Protocol v2.5.8

| type                  | command | MIDI channel | first byte          | second byte     | support              |
| --------------------- | ------- | ------------ | ------------------- | --------------- | -------------------- |
| analog I/O message    | 0xE0    | pin #        | LSB(bits 0-6)       | MSB(bits 7-13)  |          ✅          |
| digital I/O message   | 0x90    | port         | LSB(bits 0-6)       | MSB(bits 7-13)  |          ✅          |
| report analog pin     | 0xC0    | pin #        | disable/enable(0/1) | - n/a -         |          ✅          |
| report digital port   | 0xD0    | port         | disable/enable(0/1) | - n/a -         |          ✅          |
|                       |         |              |                     |                 |                      |
| start sysex           | 0xF0    |              |                     |                 |          ✅          |
| set pin mode(I/O)     | 0xF4    |              | pin # (0-127)       | pin mode        |          ✅          |
| set digital pin value | 0xF5    |              | pin # (0-127)       | pin value(0/1)  |          ✅          |
| sysex end             | 0xF7    |              |                     |                 |          ✅          |
| protocol version      | 0xF9    |              | major version       | minor version   |          ✅          |
| system reset          | 0xFF    |              |                     |                 |          ✅          |

Sysex-based sub-commands (0x00 - 0x7F) are used for an extended command set.

| type                  | sub-command | first byte       | second byte     | ...             | support             |
| --------------------- | -------     | ---------------  | --------------- | --------------- | --------------------|
| string                | 0x71        | char *string ... |                 |                 |          ✅          |
| firmware name/version | 0x79        | major version    | minor version   | char *name ...  |          ✅          |
| I2C command           | 0x76        | See [I2C](#Firmata-I2C-SYSEX-Commands)    | See [I2C](#Firmata-I2C-SYSEX-Commands)   | See [I2C](#Firmata-I2C-SYSEX-Commands)  |          ✅          |
| balena subcommand     | 0x0B        | subcommand       | see subcommands | see subcommands |          ✅          |

### Firmata SYSEX I2C Commands

I2C Read/Write request and replies are implemented following the [official protocol](https://github.com/firmata/protocol/blob/master/i2c.md).

#### I2C Config

Optional `Delay` is for I2C devices that require a delay between when the register is written to and the data in that register can be read.

Optional `I2C Mode` allows for changing the I2C interface between the balenaFin's internal pins (`SCL 18` & `SDA 19`) and the external interface (`SCL 12` & `SDA 10`). By default I2C is set to the external interface, upon initialisation under Firmata. Setting the mode to `0` will toggle the internal interface, `1` will set it to the external interface.

```
0  START_SYSEX (0xF0)
1  I2C_CONFIG (0x78)
2  Delay in microseconds (LSB) [optional]
3  Delay in microseconds (MSB) [optional]
4  I2C Mode (LSB) [optional]
5  I2C Mode (MSB) [optional]
n  END_SYSEX (0xF7)
```

### Firmata SYSEX Balena Commands

Balena SYSEX subcommands are structured under the Firmata SYSEX command. 

For example, a balena subcommand to report balena firmata firmware would be represented as follows:

`[0xF0, 0x0B, 0x00, 0xF7]`

Which represents:

`[START_SYSEX, BALENA_SUBCOMMAND, BALENA_REPORT_FIRMWARE, END_SYSEX]`

| type                   | sub-command | first byte          | second byte             | ...                     | support |
| ---------------------- | ----------- | ------------------  | ----------------------- | ----------------------- | ------- |
| report balena firmware | 0x00        |          -          |            -            |             -           |  ✅     |
| power down             | 0x01        | uint8_t init_delay  | uint8_t sleep_period[0] | uint8_t sleep_period[3] |  ✅     |
| configure IDAC         | 0x02        | uint8_t idac_mode   |      uint8_t pin        | [see idac](#idac-write-current-0x02) |  ✅     |

#### Report balena Firmware [`0x00`]

While the SYSEX command `0x79` reports the protocol version of firmata, the SYSEX balena subcommand reports the specific version of firmware that the coprocessor is running (generated by the release tag). 

#### Power Down [`0x01`]

This SYSEX command performs a hard power down of the CM3. In order to prevent loss of data or other hard shutdown consequences, users should set an `init_delay` period and gracefully power down the CM3 from the linux userspace, i.e. with `shutdown -h now`. 
After the `sleep_period` has expired, the coprocessor will resume power to the CM3 allowing it to boot into normal operating mode.

- `init_delay` is composed of **1 byte**, specified in seconds (passing 0 will immediate power down the CM3 and is not recommended!)
- `sleep_period` is composed of **4 bytes**, specified in seconds (max value of (`uint32_t` / 1000), eqv. of ~4294967 seconds)

#### IDAC Write (Current) [`0x02`]

The BGM111 has an additional Current Digital to Analogue Converter (IDAC), which can be used as a current source/sink.
This is mapped to the balena subcommand `0x02` and can be controlled in a similar way as standard `Analogue Write` command at this endpoint.
The BGM111 only has 1 IDAC peripheral so can only be used on one pin at a time.
Changing pin will remove any previous pin configurations.

Each configuration much be **selected** with `0x0F` as the MSB of that option.

| feature                 | option | first byte   | second byte   | third byte                                     | response | Note                                   |
| ------------------------| ------ | -------------| ------------- | ---------------------------------------------- | -------- | -------------------------------------- |
| Set IDAC resolution     | 0      | `0x00`       | pin # (0-127) | `0x00` to `0x03` (defaults to `0x00`)          |     -    | Will return a `SYSEX` string on error  |
| Write IDAC Pin*         | 1      | `0x01`       | pin # (0-127) | `0 - 31` (Steps)                               |     -    | Will return a `SYSEX` string on error  |
| Reset IDAC              | 2      | `0x02`       |       -       |                -                               |     -    |                                        |

> :warning: * Not all pins have access to the IDAC, see [table](#Firmata-Pin-Map)

The IDAC can be set to 4 different ranges as shown in the table below:

| Range | Min | Max | Unit |   | Step Size | Unit |
|-------|-----|-----|------|---|-----------|------|
| `0x00`|0.05 | 1.6 | uA   |   |   50      |   nA |
| `0x01`| 1.6 | 4.7 | uA   |   |   100     |   nA |
| `0x02`| 0.5 | 16  | uA   |   |   500     |   nA |
| `0x03`|  2  | 64  | uA   |   |   2       |   uA |

For examples, if you wish to select set IDAC mode on pin 6 and subsequently write an output current 8uA, it would look as follows:

`[0xF0, 0x0B, 0x02, 0x0, 0x06, 0x02]`

`[0xF0, 0x0B, 0x02, 0x1, 0x06, 0x0F]`

Which represents:

`[START_SYSEX, BALENA_SUBCOMMAND, BALENA_IDAC_WRITE, SET_RESOLUTION, PIN, RESOLUTION_MODE, END_SYSEX]`

`[START_SYSEX, BALENA_SUBCOMMAND, BALENA_IDAC_WRITE, WRITE_STEP, PIN, STEP, END_SYSEX]`

##### IDAC to DAC

The IDAC can be mapped to the same functionality as a standard DAC with additional [circuitry](https://www.silabs.com/community/mcu/8-bit/knowledge-base.entry.html/2004/03/05/generating_voltageu-PlV1).

Alternatively, `PWM` can be used to generate a psuedo-analogue output voltage.

##### Resetting IDAC

Resetting the IDAC will disable and clear the configuration on any pin currently using it.

## Firmata Pin Map

> :warning: Pins are numbered differently in the Firmata software implementation when compared to the balenaFin datasheet. This is due to the Firmata protocol, which maps pins from 0-127.

| Pin (Firmata) | Pin (Datasheet) | Port | Function              | IDAC   | Note                                                     |
|---------------|-----------------|------|-----------------------|--------|----------------------------------------------------------|
| 0             | 1               | PD14 |                       |   ✅   |                                                          |
| 1             | 4               | PB13 | SPI_CS                |   ✅   |                                                          |
| 2             | 3               | PA2  |                       |   ✅   |                                                          |
| 3             | 6               | PC8  | SPI_CLK               |        |                                                          |
| 4             | 5               | PA3  |                       |   ✅   |                                                          |
| 5             | 8               | PC6  | SPI_MOSI              |        |                                                          |
| 6             | 7               | PA4  |                       |   ✅   |                                                          |
| 7             | 10              | PC7  | SPI_MISO              |        |                                                          |
| 8             | 9               | PA5  |                       |   ✅   |                                                          |
| 9             | 12              | PA1  |                       |   ✅   |                                                          |
| 10            | 11              | PB11 | I2C_SDA (external)    |   ✅   |                                                          |
| 11            | 14              | PA0  |                       |   ✅   |                                                          |
| 12            | 13              | PF6  | I2C_SCL (external)    |        |                                                          |
| 13            | 16              | PD15 |                       |   ✅   |                                                          |
| 14            | 15              | PF7  |                       |        |                                                          |
| 15            | 18              | PD13 |                       |   ✅   |                                                          |
| 16            | PW_ON_3V3       | PF5  | PW_ON_3V3             |        | balenaFin Power Rail (used for sleep mode)               |
| 17            | PW_ON_5V        | PC9  | PW_ON_5V              |        |                                                          |
| 18            | SW_I2C_SDA_ON   | PC10 | I2C_SDA (internal)    |        | multi-master on Compute Module's I2C RTC and RGB LED     |
| 19            | SW_I2C_SCL_ON   | PC11 | I2C_SCL (internal)    |        |                                                          |

## Currently Unsupported

- [ ] SPI Support



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

# v3.0.4
## (2022-11-16)

* Switch from balenaCI to flowzone [Heath Raftery]

# v3.0.3
## (2021-04-06)

* Added clarity for Firmata pin numbering in docs. [Alex Bucknall]

# v3.0.2
## (2020-07-02)

* Corrected JS examples [Alex Bucknall]

# v3.0.1
## (2020-06-03)

* Added validation to IDAC controls [Alex Bucknall]

# v3.0.0
## (2020-06-03)

* Added IDAC Functionality [Alex Bucknall]

# v2.0.2
## (2020-05-04)

* Avoid Docker image publication [Roman Mazur]
* Updates CI to docker [Alex Bucknall]
* Adds parallel jobs to Makefile [Alex Bucknall]

# v2.0.1
## (2020-04-29)

* Fix for version tagging builds [Alex Bucknall]

# v2.0.0
## (2020-04-29)

* Added Dockerfile to build process [Alex Bucknall]

# v1.1.1
## (2020-04-27)

* Readme.md Fix [Alex Bucknall]

# v1.1.0
## (2020-04-27)

* I2C Read and Write Added to Implementation [Alex Bucknall]

# v1.0.0
## (2019-05-08)

* Update StandardFirmata.cpp [Alex Bucknall]
* Initial Beta Release [Alex Bucknall]


firmata/inc/Firmata.h
/*
  Firmata.h - Firmata library v2.5.8 - 2018-04-15
  Copyright (c) 2006-2008 Hans-Christoph Steiner.  All rights reserved.
  Copyright (C) 2009-2017 Jeff Hoefs.  All rights reserved.

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  See file LICENSE.txt for further informations on licensing terms.
*/

#ifndef Firmata_h
#define Firmata_h

#include "Boards.h"  /* Hardware Abstraction Layer + Wiring/Arduino */
#include "Serial.h"
#include "balena.h"
#include "FirmataDefines.h"
#include "FirmataMarshaller.h"
#include "FirmataParser.h"

#include <string.h>
#include <stdlib.h>

typedef unsigned char byte;

/* DEPRECATED as of Firmata v2.5.1. As of 2.5.1 there are separate version numbers for
 * the protocol version and the firmware version.
 */
#define FIRMATA_MAJOR_VERSION           2 // same as FIRMATA_PROTOCOL_MAJOR_VERSION
#define FIRMATA_MINOR_VERSION           5 // same as FIRMATA_PROTOCOL_MINOR_VERSION
#define FIRMATA_BUGFIX_VERSION          1 // same as FIRMATA_PROTOCOL_BUGFIX_VERSION

// extended command set using sysex (0-127/0x00-0x7F)
/* 0x00-0x0F reserved for user-defined commands */
// these are DEPRECATED to make the naming more consistent
#define FIRMATA_STRING          0x71 // same as STRING_DATA
#define SYSEX_I2C_REQUEST       0x76 // same as I2C_REQUEST
#define SYSEX_I2C_REPLY         0x77 // same as I2C_REPLY
#define SYSEX_SAMPLING_INTERVAL 0x7A // same as SAMPLING_INTERVAL

// pin modes
#define INPUT                   0x00 // defined in Arduino.h
#define OUTPUT                  0x01 // defined in Arduino.h
// DEPRECATED as of Firmata v2.5
#define ANALOG                  0x02 // same as PIN_MODE_ANALOG
#define PWM                     0x03 // same as PIN_MODE_PWM
#define SERVO                   0x04 // same as PIN_MODE_SERVO
#define SHIFT                   0x05 // same as PIN_MODE_SHIFT
#define I2C                     0x06 // same as PIN_MODE_I2C
#define ONEWIRE                 0x07 // same as PIN_MODE_ONEWIRE
#define STEPPER                 0x08 // same as PIN_MODE_STEPPER
#define ENCODER                 0x09 // same as PIN_MODE_ENCODER
#define IGNORE                  0x7F // same as PIN_MODE_IGNORE

namespace firmata {
// TODO make it a subclass of a generic Serial/Stream base class
class FirmataClass
{

  public:
    typedef void (*callbackFunction)(uint8_t, int);
    typedef void (*systemCallbackFunction)(void);
    typedef void (*stringCallbackFunction)(char *);
    typedef void (*sysexCallbackFunction)(uint8_t command, uint8_t argc, uint8_t *argv);

    FirmataClass();

    /* Arduino constructors */
    void begin();
    void begin(long);
    void begin(SerialClass &s);

    /* querying functions */
    void printVersion(void);
    void blinkVersion(void);
    void printFirmwareVersion(void);

    //void setFirmwareVersion(byte major, byte minor);  // see macro below
    void setFirmwareNameAndVersion(const char *name, uint8_t major, uint8_t minor);
    void disableBlinkVersion();

    /* serial receive handling */
    int available(void);
    void processInput(void);
    void parse(unsigned char value);
    bool isParsingMessage(void);

    /* serial send handling */
    void sendAnalog(byte pin, int value);
    void sendDigital(byte pin, int value); // TODO implement this
    void sendDigitalPort(byte portNumber, int portData);
    void sendString(const char *string);
    void sendString(byte command, const char *string);
    void sendSysex(byte command, byte bytec, byte *bytev);
    void write(byte c);

    /* attach & detach callback functions to messages */
    void attach(uint8_t command, callbackFunction newFunction);
    void attach(uint8_t command, systemCallbackFunction newFunction);
    void attach(uint8_t command, stringCallbackFunction newFunction);
    void attach(uint8_t command, sysexCallbackFunction newFunction);
    void detach(uint8_t command);

    /* access pin state and config */
    byte getPinMode(byte pin);
    void setPinMode(byte pin, byte config);

    /* access pin state */
    int getPinState(byte pin);
    void setPinState(byte pin, int state);

    /* utility methods */
    void sendValueAsTwo7bitBytes(int value);
    void startSysex(void);
    void endSysex(void);

  private:

    uint8_t parserBuffer[MAX_DATA_BYTES];
    FirmataMarshaller marshaller;
    FirmataParser parser;
    SerialClass *FirmataStream;

    /* firmware name and version */
    byte firmwareVersionCount;
    byte *firmwareVersionVector;

    /* pin configuration */
    byte pinConfig[TOTAL_PINS];
    int pinState[TOTAL_PINS];

    bool blinkVersionDisabled;

    /* private methods ------------------------------ */
    void strobeBlinkPin(byte pin, int count, int onInterval, int offInterval);
    friend void FirmataMarshaller::encodeByteStream (size_t bytec, uint8_t * bytev, size_t max_bytes = 0) const;

    /* callback functions */
    static callbackFunction currentAnalogCallback;
    static callbackFunction currentDigitalCallback;
    static callbackFunction currentPinModeCallback;
    static callbackFunction currentPinValueCallback;
    static callbackFunction currentReportAnalogCallback;
    static callbackFunction currentReportDigitalCallback;
    static stringCallbackFunction currentStringCallback;
    static sysexCallbackFunction currentSysexCallback;
    static systemCallbackFunction currentSystemResetCallback;

    /* static callbacks */
    inline static void staticAnalogCallback (void *, uint8_t command, uint16_t value) { if ( currentAnalogCallback ) { currentAnalogCallback(command,(int)value); } }
    inline static void staticDigitalCallback (void *, uint8_t command, uint16_t value) { if ( currentDigitalCallback ) { currentDigitalCallback(command, (int)value); } }
    inline static void staticPinModeCallback (void *, uint8_t command, uint16_t value) { if ( currentPinModeCallback ) { currentPinModeCallback(command, (int)value); } }
    inline static void staticPinValueCallback (void *, uint8_t command, uint16_t value) { if ( currentPinValueCallback ) { currentPinValueCallback(command, (int)value); } }
    inline static void staticReportAnalogCallback (void *, uint8_t command, uint16_t value) { if ( currentReportAnalogCallback ) { currentReportAnalogCallback(command, (int)value); } }
    inline static void staticReportDigitalCallback (void *, uint8_t command, uint16_t value) { if ( currentReportDigitalCallback ) { currentReportDigitalCallback(command, (int)value); } }
    inline static void staticStringCallback (void *, const char * c_str) { if ( currentStringCallback ) { currentStringCallback((char *)c_str); } }
    inline static void staticSysexCallback (void *, uint8_t command, size_t argc, uint8_t *argv) { if ( currentSysexCallback ) { currentSysexCallback(command, (uint8_t)argc, argv); } }
    inline static void staticReportFirmwareCallback (void * context, size_t, size_t, const char *) { if ( context ) { ((FirmataClass *)context)->printFirmwareVersion(); } }
    inline static void staticReportVersionCallback (void * context) { if ( context ) { ((FirmataClass *)context)->printVersion(); } }
    inline static void staticSystemResetCallback (void *) { if ( currentSystemResetCallback ) { currentSystemResetCallback(); } }
};

} // namespace firmata

extern "C" {
  // callback function types
  typedef firmata::FirmataClass::callbackFunction callbackFunction;
  typedef firmata::FirmataClass::systemCallbackFunction systemCallbackFunction;
  typedef firmata::FirmataClass::stringCallbackFunction stringCallbackFunction;
  typedef firmata::FirmataClass::sysexCallbackFunction sysexCallbackFunction;
}

extern firmata::FirmataClass Firmata;

/*==============================================================================
 * MACROS
 *============================================================================*/

/* shortcut for setFirmwareNameAndVersion() that uses __FILE__ to set the
 * firmware name.  It needs to be a macro so that __FILE__ is included in the
 * firmware source file rather than the library source file.
 */
#define setFirmwareVersion(x, y)   setFirmwareNameAndVersion(__FILE__, x, y)

#endif /* Firmata_h */
